### DISCUSSION TITLE
JSON Serialization in Flutter Application

### DISCUSSION THREAD

**Discussion by MrFahrenheit-55** [serverpod/serverpod#1714] [General]
Can you use the JSON serialization from Serverpod in the flutter application? For example 
```dart 
GeneratedProtocol.fromJSON(jsonObject).
``` 
If you can, how do you implement it with the required serializationManager?

**Comment by SandPod** [comment]
Hi @PhilipNell-258,

Yes, through the import of the `_client` package you gain access to the Serverpod serialization. Note that any object that you would like to use implicit serialization on needs to be defined as a Serverpod model.

By importing your generated client lib into your flutter file you can utilize the `Protocol()` singleton to access the serialization. The `Protocol` is a generated implementation of the serialization manager for your project.

There ware two way to serialize and deserialize objects, either you can encode the object type as part of the serialization or can manually ensure that the correct deserialization is made.

```
import 'package:<you_project_name>_client/<you_project_name>_client.dart';

void main() {
    final example = Example(name: 'alex', data: 1337);

    String withEncodedType = Protocol().encodeWithType(example);
    final decodedWithType = Protocol().decodeWithType(withEncodedType) as Example;
    
    assert(example.name == decodedWithType.name);
    assert(example.data == decodedWithType.data);

    String withoutEncodedType = SerializationManager.encode(example);
    final decodedWithoutType = Protocol().decode<Example>(withoutEncodedType);

    assert(example.name == decodedWithoutType.name);
    assert(example.data == decodedWithoutType.data);

    final fromObject = Example.fromJson(example.toJson(), Protocol());

    assert(example.name == fromObject.name);
    assert(example.data == fromObject.data);
}
```

As a last example I included use of the `fromJson method` where you can pass in any a map directly as illustrated in your question. Note that the fields of the model should then be directly accessible in the object.

**Comment by MrFahrenheit-55** [comment]
Great! Thanks @SandPod for the example code.
